### ISSUE TITLE
Query Orchestration History

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#16]
## Summary
The HTTP status query endpoint allows the caller to fetch the current execution status, input, output, and last updated time of a particular orchestration instance. However, there is no way to view the actual execution history of the instance - e.g. which activity functions it called into.

The HTTP status query endpoint should support fetching this history information and formatting it in a useful way. This history can be thought of as the "call stack" of the orchestration function and contain data such as the activities that were called, their relative order, their outputs (or errors), timers, external events received, etc.

## Motivation
There are several use cases for this kind of feature but the primary cases are likely auditing and debugging failed or "stuck" orchestrations. In the latter case, an orchestrator might be "stuck" waiting for some external event that never arrived. It would be useful for an engineer to see which event it is currently waiting on as part of this execution history.

## Design Notes
This should be built into the existing HTTP status query API handler. Because the response could be potentially very large for large orchestrations, it should only be included in the response when explicitly asked for via a query string parameter, e.g. `?showHistory=true`.

Most of the history information is available in the history table and is surfaced up via DTFx APIs. Some aggregation of this history data will be required to create a more nicely formatted output.

**Comment by SimonLuckenuik** [comment]
That would help troubleshooting instead of relying on the Table Storage directly (and create a dependency on internal implementation).

**Comment by cgillum** [comment]
Yes, absolutely.

**Comment by gled4er** [comment]
Hello @cgillum ,

I wanted to get your opinion before I start the implementation. 

I think the following format will be a good starting point for the query orchestration history:

```javascript
{
    "createdTime": "2018-01-23T05:26:37Z",
    "input": null,
    "lastUpdatedTime": "2018-01-23T05:26:58Z",
    "output": [
        "Hello Tokyo!",
        "Hello Seattle!",
        "Hello London!"
    ],
    "runtimeStatus": "Completed",
    "historyEvents": [{
        "name": "E1_SayHello",
        "scheduleId": 0, 
        "type": "TaskCompleted",
        "output": "Hello Tokyo!"
      },
      {
        "name": "E1_SayHello",
        "scheduleId": 1,
        "type": "TaskCompleted",
        "output": "Hello Seattle!"
      },
      {
        "name": "E1_SayHello",
        "scheduleId": 2,
        "type": "TaskCompleted",
        "output":  "Hello London!"
      },
      {
        "type": "TimerCreated",
        "fireAt":  "2018-01-23T06:45:38.724Z"
      },
      {
        "name": "SmsChallengeResponse",
        "type": "EventRaised",
        "input": "6925"
      },
      {
        "type": "TimerFired",
        "fireAt":  "2018-01-23T06:45:38.724Z"
      },
      {
        "type": "ExecutionTerminated"
      }]
}
```

I have also few questions:

1. Shall we try to detect and hide sensitive data such as emails, telephones and any other personal information or this is not a concern? 

2. Shall we add timestamp to each object in the historyEvent list? I think it will be better to try to order the items and show only the relevant information per event type.

3. You mentioned that we should use aggregation to reduce the contents. I think it will be good to show only TaskCompleted if we have also record for TaskScheduled for example.

Thank you!

**Comment by cgillum** [comment]
Kanio, I think this looks very good. To answer your questions:
1. I don't see a need to hide sensitive data. However, we should consider adding another flag that controls whether to return function inputs and outputs, which could potentially grow very large.
2. I think we should include timestamps. These will help provide important context about when certain actions were executed.
3. I agree - this is essentially what I had in mind.

Do you think `scheduledId` provides any value in these results? I'm wondering if we can leave that out as well.

**Comment by gled4er** [comment]
@cgillum Thank you for the quick response!

I agree with you about `scheduledId` and I will remove it from the response. 

Thank you!

**Comment by gled4er** [comment]
Hello @cgillum ,

I wanted to confirm several more things. Durable Functions are using the previous version of Durable Task Framework available here - https://github.com/Azure/durabletask/tree/rel/v1/ , right?

I wanted to clarify all use cases so I checked the various event types here - https://github.com/Azure/durabletask/blob/rel/v1/Framework/History/EventType.cs

And based on the properties I am suggesting the following approach:

| Events                | Code | Fields| History Output (default)  | History Output (Input / Output) |
| ------------------|-------|-------|--------------------|--------------------|  
| HistoryEvent      | **Base**  | EventId, **IsPlayed**, Timestamp, EventType | EventId, Timestamp, EventType |  EventId, Timestamp, EventType  
|ExecutionStarted| 0        | OrchestrationInstance*, ParentInstance*, Name, **Version**, Input, Tags| OrchestrationInstance*, ParentInstance*, Name, Tags |  OrchestrationInstance*, ParentInstance*, Name, Input, Tags
| ExecutionCompleted | 1 | OrchestrationStatus,  Result | OrchestrationStatus | OrchestrationStatus,  Result
| ExecutionFailed | 2 | N/A | N/A | N/A
| ExecutionTerminated | 3 | Input |  | Input
| TaskScheduled | 4 | Name, **Version**, Input | Name | Name, Input
| TaskCompleted | 5 | **TaskScheduledId**, Result | | Result 
| TaskFailed | 6 | **TaskScheduledId**, Reason, Details | Reason, Details | Reason, Details
| SubOrchestrationInstanceCreated | 7 | Name, **Version**, InstanceId, Input | Name,InstanceId | Name, InstanceId, Input
| SubOrchestrationInstanceCompleted | 8 | **TaskScheduledId**, Result | Result | Result 
| SubOrchestrationInstanceFailed | 9 |  **TaskScheduledId**, Reason, Details | Reason, Details | Reason, Details
| TimerCreated | 10 | FireAt | FireAt	| FireAt
| TimerFired | 11 | TimerId | TimerId | TimerId	
| OrchestratorStarted | 12 | | |
| OrchestratorCompleted | 13 | | |
| EventRaised  | 14 | Name, Input | Input | Name, Input
| ContinueAsNew | 15 | | | 
| GenericEvent	| 16 | Data | Data | Data

So with this approach we are removing from the response  **IsPlayed**, **Version**,  **TaskScheduledId** and Input / Result if they are not requested explicitly. 	

In terms of the aggregation we will skip the event types on the left for the event type on the right (if both the events on right and left are present in the execution history):

-  TaskScheduled	  -> 	TaskCompleted / TaskFailed		
-  ExecutionStarted	  ->   ExecutionCompleted / ExecutionFailed / ExecutionTerminated			
-  SubOrchestrationInstanceCreated -> SubOrchestrationInstanceCompleted / SubOrchestrationInstanceFailed	
-  TimerCreated	   ->  TimerFired	
- OrchestratorStarted ->  OrchestratorCompleted	

Let me know if this is an acceptable approach.

Thank you!

**Comment by cgillum** [comment]
The high-level approach seems good to me, we can work out the details as we look at examples from your implementation (I don't think things like `HistoryEvent.EventId` will be useful, for example).

I'm debating with myself about TaskScheduled. On one hand, it's nice to filter it out to reduce the output size. On the other hand, it's useful to know _when_ a task was scheduled. One potential compromise would be to add a `ScheduledTime` field to `TaskCompleted` so that users can get an idea for the latency of a particular activity function. Thoughts?

**Comment by cgillum** [comment]
This has been merged into **dev** and will go out in the next release.
